File: Quick Union/QuickUnion.py
class QuickUnion:
    def __init__(self, N):
        self.p = [k for k in range(N)]
    
    def find(self, i):
        while (i != self.p[i]):
            i = self.p[i]
        return i

    def union(self, i, j):
        ri = self.find(i)
        rj = self.find(j)
        if (ri != rj):
            self.p[ri] = rj

N = int(input())
M = int(input())
lines = [input().split() for _ in range(M)]
QU = QuickUnion(N)
for line in lines:
    if line[0]=='F':
        print(QU.find(int(line[1])))
    elif line[0]=='U':
        QU.union(int(line[1]),int(line[2]))

File: Quick Union/QuickUnion.cpp
#include <bits/stdc++.h>
using namespace std;

class QuickUnion{
    vector<int> p;

    public:
        QuickUnion(int N){
            p = vector<int>(N);
            for (int i=0;i<N;i++){
                p[i] = i;
            }
        }

        int find(int i) {
            while (i != p[i]){
                i = p[i];
            }
            return i;
        }

        void qunion(int i, int j){
            int ri = find(i);
            int rj = find(j);
            if (ri != rj) 
                p[ri] = rj;
        }
};

int main() {
    int N,M;
    cin>>N;
    cin>>M;
    QuickUnion qu = QuickUnion(N);
    for (int i=0;i<M;i++){
        char c;
        cin>>c;
        if (c=='F'){
            int x;
            cin>>x;
            cout<<qu.find(x)<<endl;
        } else if (c=='U') {
            int x,y;
            cin>>x>>y;
            qu.qunion(x,y);
        }
    }
}
